core/controllers/acl_decorators.py

```diff
@@ -2933,3 +2933,85 @@ def test_request_originates_from_valid_oppia_ml_instance(self, **kwargs):
     test_request_originates_from_valid_oppia_ml_instance.__wrapped__ = True
 
     return test_request_originates_from_valid_oppia_ml_instance
+
+
+def can_update_suggestion(handler):
+    """Decorator to check whether the current user can update suggestions.
+
+    Args:
+        handler: function. The function to be decorated.
+
+    Returns:
+        function. The newly decorated function that now checks
+        if the user can update a given suggestion.
+
+    Raises:
+        NotLoggedInException. The user is not logged in.
+        UnauthorizedUserException. The user does not have credentials to
+            edit this suggestion.
+        InvalidInputException. The submitted suggestion id is not valid.
+        PageNotFoundException. A suggestion is not found with the given
+            suggestion id.
+    """
+    def test_can_update_suggestion(
+            self, suggestion_id, **kwargs):
+        """Returns a handler to test whether a suggestion can be updated based
+        on the user's roles.
+
+        Args:
+            suggestion_id: str. The suggestion id.
+            **kwargs: *. Keyword arguments.
+
+        Returns:
+            function. The handler for updating a suggestion.
+
+        Raises:
+            NotLoggedInException. The user is not logged in.
+            UnauthorizedUserException. The user does not have credentials to
+                edit this suggestion.
+            InvalidInputException. The submitted suggestion id is not valid.
+            PageNotFoundException. A suggestion is not found with the given
+                suggestion id.
+        """
+        if not self.user_id:
+            raise base.UserFacingExceptions.NotLoggedInException
+        user_actions = self.user.actions
+
+        if len(suggestion_id.split('.')) != 3:
+            raise self.InvalidInputException(
+                'Invalid format for suggestion_id.' +
+                ' It must contain 3 parts separated by \'.\'')
+
+        suggestion = suggestion_services.get_suggestion_by_id(suggestion_id)
+
+        if suggestion is None:
+            raise self.PageNotFoundException
+
+        if role_services.ACTION_ACCEPT_ANY_SUGGESTION in user_actions:
+            return handler(self, suggestion_id, **kwargs)
+
+        if suggestion.author_id == self.user_id:
+            raise base.UserFacingExceptions.UnauthorizedUserException(
+                'You are not allowed to update suggestions that you created.')
+
+        if suggestion.suggestion_type not in (
+                feconf.CONTRIBUTOR_DASHBOARD_SUGGESTION_TYPES):
+            raise self.InvalidInputException('Invalid suggestion type.')
+
+        else:
```

Comment:
You can probably drop this "else" and deindent all the stuff below by 4.

Reply:
Done